rfcs/2020-06-18-2625-architecture-revisit.md

```diff
@@ -0,0 +1,504 @@
+   of the mental model needed to understand Vector's behavior.
+
+2. It's possible that tying computational concurrency to source task concurrency
+   results will work well in some situations and work poorly in others. It adds
+   a significant level of variability to something that used to be relatively
+   consistent. This applies to both performance as well as memory use.
+
+3. Many of the benefits here will not come automatically. We'll need to make
+   sure that we keep the overall design in mind and make individual components
+   that work well with it. This applies particularly to sources that don't
+   currently run more than one task (e.g. file source, UDP-based sources, etc).
+
+4. It's not entirely clear how runtimes like Lua and WASM should fit into this
+   model. Their biggest benefit is the flexibility to do anything, which goes
+   against our strategy of limiting transform capability. They also have
+   relatively unknown memory requirements, which could make them unsuitable to
+   run potentially thousands of copies of in a high-concurrency server scenario.
+   The best solution here may be something like the Lua transform's hooks, which
+   we can inspect at runtime to determine the required capabilities.
+
+## Rationale and Alternatives
+
+The rationale for these changes is to give us the maximum amount of future
+flexibility for the least amount of work. If we can isolate as much code as
+possible from the effects of future changes, we should be in a good position to
+move forward smoothly.
+
+The biggest alternative to doing this kind of reworking would be to stick with
+the existing overall model and try to work things like concurrency into each
+individual component. One example proposal along these lines was to have
+transforms batch up events and spawn a new task to process them. This would
+likely require some degree of similar changes in terms of defining types of
+transforms, but would keep topology-level architecture untouched.
+
+Making fewer, less invasive changes would have the strong benefit of being
+cheaper, but I worry that the improvements we'd be able to make with such
+a limited view (i.e. only within a single component) would keep us rather
+limited.
+
+When considering specifically the change to push transforms into source tasks,
+there are a number of alternatives for getting that kind of concurrency. While
+I believe that design has the most promise, one benefit of this proposal is that
+it becomes an implementation detail. The other changes around component and
+topology design should give us the freedom to experiment with different ways of
+achieving concurrency under the hood.
+
+## Outstanding Questions
+
+1. Where do runtimes fit into the new model of transforms?
+
+2. Should we (and if so, how should we) add concurrency to components like the
+   file source and UDP-based socket and syslog?
+
+3. Should we limit the total amount of concurrency somehow? Is that possible
```

Comment:
I don't think so for the reason you alluded to. It does not seem possible. If we did want to limit concurrency we should limit the higher-level concurrency factor, like the number of connections maintained.